src/main/scala/cromwell/engine/SymbolTable.scala

```diff
@@ -0,0 +1,58 @@
+package cromwell.engine
+
+import cromwell.binding.types.WdlType
+import cromwell.binding.{FullyQualifiedName, WdlBinding, WdlValue}
+
+import scala.collection.mutable
+
+case class SymbolTableEntry(scope: String, name: String, iteration: Option[Int], wdlValue: Option[WdlValue], wdlType: WdlType, input: Boolean) {
+
+  wdlValue match {
+    case Some (value) => wdlType.checkCompatible(value.value)
+    case None => // Nothing is always okay for a value.
+  }
+
+  def isInput: Boolean = input
+
+  def isOutput: Boolean = !isInput
+
+  /**
+   * Create a copy of the receiver with `wdlValue` set to `newValue`.
+   */
+  def copyWithUpdatedValue(newValue: WdlValue): SymbolTableEntry =
+    SymbolTableEntry(scope, name, iteration, Option(newValue), wdlType, input)
+}
+
+class SymbolTable(binding: WdlBinding, inputs: Map[FullyQualifiedName, WdlValue]) {
+  val table = mutable.Set[SymbolTableEntry]()
+
+  binding.workflow.calls.foreach { call =>
+    call.task.inputs.foreach { case (inputName, wdlType) =>
+      table += new SymbolTableEntry(call.fullyQualifiedName, inputName, iteration = None, wdlValue = None, wdlType, input = true)
+    }
+    call.task.outputs.foreach { output =>
+      table += new SymbolTableEntry(call.fullyQualifiedName, output.name, iteration = None, wdlValue = None, output.wdlType, input = false)
+    }
+  }
+
+  private def assignSymbolTableEntry(fullyQualifiedName: FullyQualifiedName, value: WdlValue, iteration: Option[Int], input: Boolean): Unit = {
+    val lastIndex = fullyQualifiedName.lastIndexOf(".")
+    val (scope, name) = (fullyQualifiedName.substring(0, lastIndex), fullyQualifiedName.substring(lastIndex + 1))
+
+    val maybeOldEntry = table.find { entry =>
```

Comment:
Actually it's pretty funky, not sure if it's neat. What you're doing is glomming on some extra crap to the ADT of the case class, but it's not really part of the case class for any definition of a case class that a normal human would have.